subscription.go

```diff
@@ -5,66 +5,161 @@ import (
+	// ContextValue    context.Context
 	VariableValues  map[string]interface{}
 	OperationName   string
 	FieldResolver   FieldResolveFn
 	FieldSubscriber FieldResolveFn
 }
 
-// Subscribe performs a subscribe operation
-func Subscribe(ctx context.Context, p SubscribeParams) chan *Result {
-	resultChannel := make(chan *Result)
+// SubscriptableSchema implements `graphql-transport-ws` `GraphQLService` interface: https://github.com/graph-gophers/graphql-transport-ws/blob/40c0484322990a129cac2f2d2763c3315230280c/graphqlws/internal/connection/connection.go#L53
+// you can pass `SubscriptableSchema` to `graphql-transport-ws` `NewHandlerFunc`
+type SubscriptableSchema struct {
+	Schema     Schema
+	RootObject map[string]interface{}
+}
+
+// Subscribe method let you use SubscriptableSchema with graphql-transport-ws https://github.com/graph-gophers/graphql-transport-ws
+func (self *SubscriptableSchema) Subscribe(ctx context.Context, queryString string, operationName string, variables map[string]interface{}) (<-chan interface{}, error) {
+	c := Subscribe(Params{
+		Schema:         self.Schema,
+		Context:        ctx,
+		OperationName:  operationName,
+		RequestString:  queryString,
+		RootObject:     self.RootObject,
+		VariableValues: variables,
+	})
+	to := make(chan interface{})
+	go func() {
+		defer close(to)
+		select {
+		case <-ctx.Done():
+			return
+		case res, more := <-c:
+			if !more {
+				return
+			}
+			to <- res
+		}
+	}()
+	return to, nil
+}
+
+// Subscribe performs a subscribe operation on the given query and schema
+// currently does not support extensions hooks
+func Subscribe(p Params) chan *Result {
+
+	source := source.NewSource(&source.Source{
+		Body: []byte(p.RequestString),
+		Name: "GraphQL request",
+	})
+
+	// TODO run extensions hooks
+
+	// parse the source
+	AST, err := parser.Parse(parser.ParseParams{Source: source})
+	if err != nil {
+
+		// merge the errors from extensions and the original error from parser
+		return sendOneResultAndClose(&Result{
+			Errors: gqlerrors.FormatErrors(err),
+		})
+	}
+
+	// validate document
+	validationResult := ValidateDocument(&p.Schema, AST, nil)
+
+	if !validationResult.IsValid {
+		// run validation finish functions for extensions
+		return sendOneResultAndClose(&Result{
+			Errors: validationResult.Errors,
+		})
+
+	}
+	return ExecuteSubscription(ExecuteParams{
+		Schema:        p.Schema,
+		Root:          p.RootObject,
+		AST:           AST,
+		OperationName: p.OperationName,
+		Args:          p.VariableValues,
+		Context:       p.Context,
+	})
+}
+
+func sendOneResultAndClose(res *Result) chan *Result {
+	resultChannel := make(chan *Result, 1)
+	resultChannel <- res
+	close(resultChannel)
+	return resultChannel
+}
+
+// ExecuteSubscription is similar to graphql.Execute but returns a channel instead of a Result
+// currently does not support extensions
+func ExecuteSubscription(p ExecuteParams) chan *Result {
+
+	if p.Context == nil {
+		p.Context = context.Background()
```

Comment:
I'm ok with this since it looks like you can still stop the subscription by closing the channel.

Reply:
yes, to finish a sub you can close the channel from inside the resolver
If you want to end a sub from outside instead you cancel the context